guides/keras_cv/object_detection_keras_cv.py

```diff
@@ -0,0 +1,872 @@
+bounding boxes, and must update them accordingly.
+
+Luckily, KerasCV natively supports bounding box augmentation with its extensive
+library
+of [data augmentation layers](https://keras.io/api/keras_cv/layers/preprocessing/).
+The code below loads the Pascal VOC dataset, and performs on-the-fly bounding box
+friendly data augmentation inside of a `tf.data` pipeline.
+"""
+augmenter = keras.Sequential(
+    layers=[
+        keras_cv.layers.RandomFlip(mode="horizontal", bounding_box_format="xywh"),
+        keras_cv.layers.JitteredResize(
+            target_size=(640, 640), scale_factor=(0.75, 1.3), bounding_box_format="xywh"
+        ),
+    ]
+)
+
+train_ds = train_ds.map(augmenter, num_parallel_calls=tf.data.AUTOTUNE)
+visualize_dataset(
+    train_ds, bounding_box_format="xywh", value_range=(0, 255), rows=2, cols=2
+)
+
+"""
+Great!  We now have a bounding box friendly data augmentation pipeline.
+Let's format our evaluation dataset to match.  Instead of using
+`JitteredResize`, let's use the deterministic `keras_cv.layers.Resizing()`
+layer.
+"""
+
+inference_resizing = keras_cv.layers.Resizing(
+    640, 640, bounding_box_format="xywh", pad_to_aspect_ratio=True
+)
+eval_ds = eval_ds.map(inference_resizing, num_parallel_calls=tf.data.AUTOTUNE)
+
+"""
+Due to the fact that the resize operation differs between the train dataset,
+which uses `JitteredResize()` to resize images, and the inference dataset, which
+uses `layers.Resizing(pad_to_aspect_ratio=True)`. it is good practice to
+visualize both datasets:
+"""
+
+visualize_dataset(
+    eval_ds, bounding_box_format="xywh", value_range=(0, 255), rows=2, cols=2
+)
+
+"""
+Finally, let's unpackage our inputs from the preprocessing dictionary, and
+prepare to feed the inputs into our model.  In order to be TPU compatible,
+bounding box Tensors need to be `Dense` instead of `Ragged`.  If training on
+GPU, you can omit the `bounding_box.to_dense()` call.  If ommitted,
+the KerasCV RetinaNet
+label encoder will automatically correctly encode Ragged training targets.
+"""
+
+
+def dict_to_tuple(inputs):
+    return inputs["images"], bounding_box.to_dense(
+        inputs["bounding_boxes"], max_boxes=32
+    )
+
+
+train_ds = train_ds.map(dict_to_tuple, num_parallel_calls=tf.data.AUTOTUNE)
+eval_ds = eval_ds.map(dict_to_tuple, num_parallel_calls=tf.data.AUTOTUNE)
+
+train_ds = train_ds.prefetch(tf.data.AUTOTUNE)
+eval_ds = eval_ds.prefetch(tf.data.AUTOTUNE)
+
+"""
+
+### Optimizer
+
+In this guide, we use a standard SGD optimizer and rely on the
+[`keras.callbacks.ReduceLROnPlateau`](https://keras.io/api/callbacks/reduce_lr_on_plateau/)
+callback to reduce the learning rate.
+
+You will always want to include a `global_clipnorm` when training object
+detection models.  This is to remedy exploding gradient problems that frequently
+occur when training object detection models.
+"""
+
+base_lr = 0.005
+# including a global_clipnorm is extremely important in object detection tasks
```

Comment:
In my experience, a global clipnorm will prevent training runs from going to NaNs, but they will still be bad. This will show as bumpy training curves. Essentially, the clip norm puts the training back on track when it diverges but the training epochs up to that point are lost. Once you have found correct LR settings, global clipnorm is no longer necessary. You can look at all the "arthropods" samples, even the 32-core TPU pod one using a batch size of 1024. They all converge nicely and none use global clipnorm.